foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_10_0/PomDependencyVersionMigration.java

```diff
@@ -0,0 +1,147 @@
+            String updatedVersion = getUpdatedDependencyVersion(matchingDependency);
+
+            // Replace the version in place
+            if (matchingDependency.getVersion() != null) {
+                modifications.add(replaceInTag(matchingDependency, "version", updatedVersion));
+            } 
+            // Add a new version tag
+            else {
+                int indentSize = matchingDependency.getLocation("artifactId" + START).getColumnNumber() - 1;
+                // assumes spaces instead of tabs, but accounting for tabs would be more trouble than it's worth IMO
+                String indent = StringUtils.repeat(' ', indentSize);
+                modifications.add(new PomModifications.Insertion(matchingDependency.getLocation(END), 0,
+                        ignore -> indent + "<version>" + updatedVersion + "</version>\n"));
+            }
+        }
+
+        if (!modifications.isEmpty()) {
+            PomHelper.writeModifications(pomFile, modifications.finalizeMods());
+        }
+        return true;
+    }
+
+    /**
+     * Creates a list of {@link Dependency}'s from their string representations: {group id}:{artifact id}:{version}
+     */
+    private List<Dependency> createDependencies() {
+        String[] dependencyStrings = {
+            "org.scala-lang:scala-reflect:${version.scala}",
+            "org.apache.commons:commons-math3:${version.commons.math}",
+            "io.smallrye.reactive:smallrye-reactive-messaging-kafka:${version.smallrye.reactive.messaging}",
+            "io.smallrye.reactive:smallrye-reactive-messaging-in-memory:${version.smallrye.reactive.messaging}",
+            "junit:junit:${version.junit}",
+            "org.jboss.resteasy:resteasy-client:${version.resteasy}",
+            "org.jboss.resteasy:resteasy-jackson2-provider:${version.resteasy}",
+            "org.awaitility:awaitility:${version.awaitility}",
+            "io.vertx:vertx-core:${version.vertx}",
+            "io.smallrye.config:smallrye-config:${version.smallrye.config}",
+            "org.slf4j:slf4j-api:${version.slf4j}",
+            "org.slf4j:slf4j-simple:${version.slf4j}"
+        };
+
+        return Arrays.stream(dependencyStrings)
+            .map(dependencyString -> {
+                String[] dependencyAttributes = dependencyString.split(":");
+                Dependency dependency = new Dependency();
+                dependency.setGroupId(dependencyAttributes[0]);
+                dependency.setArtifactId(dependencyAttributes[1]);
+                dependency.setVersion(dependencyAttributes[2]);
+                return dependency;
+            })
+            .toList();
+    }
+
+    /**
+     * Returns a list of any dependencies that need to be updated with a new version
+     */
+    private List<Dependency> getMatchingDependenciesWithDifferentVersionsForProject(Model model) {
+        List<Dependency> matchingDependencies = new ArrayList<>();
+
+        this.dependenciesToUpdate.stream()
```

Comment:
A: Why not invert this, and let the project dependency collection be the "outer" stream (and handled by the parent class method). Then `isMatchingDependencyWithDifferentVersion`, could be more like `dependencyRequiresUpdate`. Then _it_ streams the `dependenciesToUpdate` looking for a match.  That would also allow you to break out the different aspects of the filter, first comparing groupId, then artifactId, and lastly version.

Reply:
Updated to stream all dependencies initially, and stream dependenciesToUpdate in the helper method